1-theory/2-different-paradigms.adoc

```diff
@@ -0,0 +1,217 @@
+
+This centralization of state into the server has benefits for systems with multiple different clients who offer similar workflows. Instead of distributing all the logic, checking data fields, showing lists of "Actions", etc. around various clients — who might come to different conclusions — REST keeps it all in one place.
+
+This book wil get more in depth on hypermedia controls later. There are a few other important things to understand about REST APIs first:
+
+- REST must be stateless: not persisting sessions between requests
+- Responses should declare cacheablility: helps your API scale if clients respect the rules
+- REST focuses on uniformity: if you’re using HTTP you should utilize HTTP features whenever possible, instead of inventing conventions
+
+These constraints of REST when applied to HTTP APIs can help the API last for decades, which is a whole lot more complex without these concepts.
+
+REST does not require the use of schema metadata (like WSDL or similar), but does allow it. In fact REST has no opinions either way: it does not explicitly demand it, nor disallow it. the metadata is something many API developers hated about SOAP, and these days it is once again all the rage in gRPC and GraphQL, but REST has a few type systems available for optional use, the main one these days being: JSON Schema.
+
+JSON Schema is inspired by XML Schema — but not functionally identical — and is one of the most important things to happen to HTTP APIs in years, and will be discussed a lot throughout the book.
+
+Unfortunately, REST become a marketing buzzword for most of 2006–2014. It became a metric of quality that developers would aspire to, fail to understand, then label as REST anyway. Most systems saying they are REST are little more than RPC with HTTP verbs and pretty URLs. As such, you might not get cacheability provided, it might have a bunch of wacky conventions, and there might not be any links for you to use to discover next available actions. These APIs are jokingly called REST__ish__ by people aware of the difference.
+
+Whilst REST is actually a series of layers of abstraction on top of RPC, it is primarily an RDA. You are interacting with resources, creating things and modifying things, so instead of "sending an email" you would "create an email" with all the information it needs and it would send once it is created. Or you would create an "send email attempt" resource. Basically instead of executing a thing, you are manipulating domain models into doing what you want, and the next available actions are provided to continue making other manipulations later. These next available actions can be entirely ignored, and clients can work with it in a standard RPC-style approach, but that is usually not a fantastic idea.
```

Comment:
```suggestion
Whilst REST is actually a series of layers of abstraction on top of RPC, it is primarily an RDA. You are interacting with resources, creating things and modifying things, so instead of "sending an email" you would "create an email" with all the information it needs and it would send once it is created. Or you would create a "send email attempt" resource. Basically instead of executing a thing, you are manipulating domain models into doing what you want, and the next available actions are provided to continue making other manipulations later. These next available actions can be entirely ignored, and clients can work with it in a standard RPC-style approach, but that is usually not a fantastic idea.
```